enterprise/admin/views.py

```diff
@@ -94,19 +97,113 @@ def post(self, request, customer_uuid):
+
+        # initial form validation - check that form data is well-formed
+        if manage_learners_form.is_valid():
+            # The form is valid. Call the appropriate helper depending on the mode:
+            mode = manage_learners_form.cleaned_data[ManageLearnersForm.Fields.MODE]
+            if mode == ManageLearnersForm.Modes.MODE_SINGULAR:
+                self._handle_singular(enterprise_customer, manage_learners_form)
+            else:
+                self._handle_bulk_upload(enterprise_customer, manage_learners_form, request)
+
+        # _handle_form might add form errors, so we check if it is still valid
         if manage_learners_form.is_valid():
-            EnterpriseCustomerUser.objects.link_user(enterprise_customer, manage_learners_form.cleaned_data["email"])
-            manage_learners_form = ManageLearnersForm()
+            # Redirect to GET if everything went smooth.
+            return HttpResponseRedirect("")
 
+        # if something went wrong - display bound form on the page
         context = self._build_context(request, customer_uuid)
-
         context.update({self.ContextParameters.MANAGE_LEARNERS_FORM: manage_learners_form})
-
         return render(request, self.template, context)
 
+    @classmethod
+    def _handle_singular(cls, enterprise_customer, manage_learners_form):
+        """
+        Link single user by email or username.
+
+        Arguments:
+            enterprise_customer (EnterpriseCustomer): learners will be linked to this Enterprise Customer instance
+            manage_learners_form (ManageLearnersForm): bound ManageLearners form instance
+        """
+        form_field_value = manage_learners_form.cleaned_data[ManageLearnersForm.Fields.EMAIL_OR_USERNAME]
+        email = email_or_username__to__email(form_field_value)
+        try:
+            validate_email_to_link(email, form_field_value, ValidationMessages.INVALID_EMAIL_OR_USERNAME)
+            EnterpriseCustomerUser.objects.link_user(enterprise_customer, email)
+        except ValidationError as exc:
+            manage_learners_form.add_error(ManageLearnersForm.Fields.EMAIL_OR_USERNAME, exc.message)
+
+    @classmethod
+    def _handle_bulk_upload(cls, enterprise_customer, manage_learners_form, request):
+        """
+        Bulk link users by email.
+
+        Arguments:
+            enterprise_customer (EnterpriseCustomer): learners will be linked to this Enterprise Customer instance
+            manage_learners_form (ManageLearnersForm): bound ManageLearners form instance
+            request (django.http.request.HttpRequest): HTTP Request instance
+        """
+        errors = []
+        emails = set()
+        already_linked_emails = []
+        duplicate_emails = []
+        csv_file = manage_learners_form.cleaned_data[ManageLearnersForm.Fields.BULK_UPLOAD]
+        try:
+            parsed_csv = parse_csv(csv_file, expected_columns={ManageLearnersForm.CsvColumns.EMAIL})
```

Comment:
@bradenmacdonald nit: I've got a slight concern that this method is now ~60 LOC - might be good to extract this part into a dedicated method.